src/almanack/metrics/data.py

```diff
@@ -770,3 +783,141 @@ def get_github_build_metrics(
+        primary_language (Optional[str]):
+            The primary programming language of the repository.
+
+    Returns:
+        Optional[dict[str,Any]]:
+            Code coverage data or an empty dictionary if unable
+            to find code coverage data.
+    """
+
+    if primary_language is None:
+        return {}
+
+    if primary_language.lower() == "python":
+        return parse_python_coverage_data(repo)
+
+    return {}
+
+
+def parse_python_coverage_data(
+    repo: pygit2.Repository,
+) -> Optional[Dict[str, Any]]:
+    """
+    Parses coverage.py data from recognized formats such as JSON, XML, or LCOV.
+
+    Args:
+        repo (pygit2.Repository):
+            The pygit2 repository object containing code.
+
+    Returns:
+        Optional[Dict[str, Any]]:
+            A dictionary with standardized code coverage data or an
+            empty dict if no data is found.
+    """
+    coverage_files = [
+        "coverage.json",
+        "coverage.xml",
+        "coverage.lcov",
+    ]
+    for coverage_file in coverage_files:
+        if (
+            coverage_object := find_file(repo=repo, filepath=coverage_file)
+        ) is not None:
+            file_path = f"{repo.workdir}/{coverage_object.name}"
+            total_lines = executed_lines = 0
+            timestamp = None
+
+            try:
+                if coverage_file.endswith(".json"):
+                    # Parse JSON coverage data
+                    with open(file_path, "r") as f:
+                        coverage_data = json.load(f)
+
+                    # Use the `summary` key directly
+                    summary = coverage_data.get("summary", {})
+                    total_lines = summary.get("num_statements", 0)
+                    executed_lines = summary.get("covered_lines", 0)
+                    coverage_percentage = summary.get("percent_covered", 0.0)
+
+                    # Retrieve the timestamp
+                    timestamp = datetime.fromisoformat(
+                        coverage_data["meta"]["timestamp"]
+                    )
+
+                elif coverage_file.endswith(".xml"):
+                    # Parse XML coverage data (using defusedxml for safety)
+                    tree = ET.parse(file_path)
+                    root = tree.getroot()
+
+                    # Extract the total lines and executed lines directly from the root element
+                    total_lines = int(root.attrib.get("lines-valid", 0))
+                    executed_lines = int(root.attrib.get("lines-covered", 0))
+
+                    # Optional: parse timestamp if present in XML attributes
+                    timestamp_str = root.attrib.get("timestamp")
+                    timestamp = datetime.fromtimestamp(
+                        float(timestamp_str)
+                        / 1000  # Convert from milliseconds to seconds
+                    )
+
+                elif coverage_file.endswith(".lcov"):
```

Comment:
It might be worth specifying what this file type is - what generated it?

Reply:
Thanks for this - `[lcov` files are one of the report types](https://coverage.readthedocs.io/en/7.6.8/cmd.html#cmd-lcov) generated by `coverage.py`. I'll add a comment about this. Generally they are a bit more compressed than that of `xml` or `json` but lack datetime stamps which is the reason for inferring that detail from the git object within this if-branch.